hw/top_earlgrey/data/pins_cw310.xdc

```diff
@@ -43,15 +43,25 @@ set_property -dict { PACKAGE_PIN W8 IOSTANDARD LVCMOS18 }  [get_ports { IOA5 }];
 set_property -dict { PACKAGE_PIN W10 IOSTANDARD LVCMOS18 } [get_ports { IOA6 }]; #USRDIP6
 set_property -dict { PACKAGE_PIN V11 IOSTANDARD LVCMOS18 } [get_ports { IOA7 }]; #USRDIP7
 
-## SPI/JTAG
+## SPI / JTAG (part of it, other JTAG signals further below)
```

Comment:
o damn, so even on the cw310 some of the spi device / jtag pins are shared? That's...great.
Even more reason for another xdc i guess.

Reply:
Yeah that is something we still need to figure out. 
I think there are enough pins on the SAM3X to disentangle them - but I first wanted to get the SW straps correctly mapped.

Reply:
yeah i think i'm just not sure what lines are actually connected between the kintex and SAM3X right now... very confused.

Reply:
I looked at the schematic and the sample XDC provided by newAE.
The SAM3X has a bunch of GPIO lines that are labeled as `USB_D[*]` and `USB_A[*]` that are presumably intended to implement some sort of simple register read/write protocol.

We do not really implement such an interface, but map some of of our strap pins to `USB_A[*]`.
I just followed this pattern here and mapped more strap pins to that interface (and rearranged them a bit).

Reference schematic and XDC files for the CW310 can be found [here](https://rtfm.newae.com/Targets/CW310%20Bergen%20Board/#reference-material-schematics).

Reply:
We'll need to ask our partners to enable moving the MPSSE interface if we want JTAG to get off of the SPI pins. Currently, there's just the one 4-wire interface, and it's either a proper SPI controller (faster) or MPSSE emulation via GPIO bit-banging (slow SPI, but makes JTAG and other protocols available).

However, this would be an even more OT-specific request, since I'm not seeing another interface that fits in with their usual operations. Perhaps we could run the MPSSE interface over the USB_SPARE set, though USB_SPARE0 being on its own bank isn't ideal--At least it's on an adjacent bank, though. They might be open to changes, though, considering the 4 MPSSE GPIOs consume upper bits of the USB_A[*] bus.

Reply:
Just looped in on changes happening - the "JTAG/SPI sharing" system was I think mostly there to make things backwards compatible as much as possible in the initial stages. We had no particular "love" for that system.

It's not an issue to move the MPSSE pins to any other unused pins (in the firmware they are all bit-banged, for reference defined here - https://github.com/newaetech/cw310-bergen-board/blob/mpsse/microfw/CW310/src/CW310_platform.h#L65). I think we could change that to run-time, we hadn't bothered as the current use-case they were always fixed due to external connections.

Ideally the SPI pins would remain the same as the SPI interface on the SAM3X is a hardware block. We *can* bit-bang SPI but obviously it runs slower, so this isn't a hard requirement if it is preferable to move them for other reasons.

The official MPSSE interface only has 4 extra GPIO pins (we implement the 8-bit mode). If there are other strapping pins it would be easier (I think) to set them via Python, especially if they are simply fixed state (OpenOCD doesn't need to know about them). You need to do Python configuration for OpenOCD mode anyway, so there is no real disadvantage to using the Python to set them to a fixed state.

So we'll need to move the MPSSE-GPIO pins to make sure they align with ones that need to be toggled for the JTAG access process (assuming that is required, if OpenOCD doesn't need to know about them less critical). The current Python/firmware combo lets you set arbitrary I/O pins on the SAM3X to given states.

Reply:
@colinoflynn Thanks! I think we'll want one of the MPSSE-GPIO pins for TRST, but in any case, all of the JTAG pins for OT are actually sharing muxed pins, so anything suitable for GPIO on the SAM3X works for us. OT's SPI device is on dedicated pins (i.e. _not_ muxed), so that would work nicely with having the SAM3X's SPI interface also be dedicated to SPI. :smile: 

For MPSSE emulation at run time, OT only needs to be able to turn it on/off, so the MPSSE pins can be used as ordinary GPIOs when JTAG is inactive. OT doesn't have a need for run-time-configurable pin mapping--I brought up dynamic configuration in case that might be useful for other applications the CW310 might target. However, if there is a fixed mapping that would work for all of us, that sounds good to me!

Would you generally choose to commandeer the most significant bits of USB_A[*] for this?